### PR TITLE
Strip Mosaic kernel debug info and exclude autotune cache dir from cache keys

### PR DESCRIPTION
**Pull request by colin-ant**
Two fixes to compilation cache keys:

1. Exclude `xla_gpu_per_fusion_autotune_cache_dir` from the cache key. it's currently auto-set to a filesystem path that varies between environments.
2. Strip debug info from Mosaic kernel bytecode. The top-level `strip-debuginfo` pass doesn't reach inside the base64-encoded kernel MLIR in `tpu_custom_call`, so kernel source paths leak into the cache key.

### CODE REVIEW

File: jax/_src/cache_key.py [hunk 1]
```diff
@@ -190,6 +192,46 @@ def _update_bc_attribute(op: ir.Operation) -> ir.WalkResult:
   return m
 
 
+def _strip_mosaic_debug_info(m: ir.Module) -> None:
```
Comment (superbobry) [inline jax/_src/cache_key.py:200]: OOC why do this as a post-processing step as opposed to via an argument to `tpu_custom_call.as_tpu_kernel` or `pltpu.CompilerParams`?

Comment (colin-ant) [reply jax/_src/cache_key.py:200]: iiuc that would require us to either:

1. re-run the lowering with a `strip_mosaic_debug_info` arg while making the cache key
2. strip debug info from the final executable (with the post-processing step, the executable still has the mosaic debug info)

(I could be misunderstanding though)

File: jax/_src/cache_key.py [hunk 2]
```diff
@@ -190,6 +192,46 @@ def _update_bc_attribute(op: ir.Operation) -> ir.WalkResult:
...
+        kernel = ir.Module.parse(base64.b64decode(body), context=ctx)
+        pm.PassManager.parse("builtin.module(strip-debuginfo)").run(
+            kernel.operation)
+        out = io.BytesIO()
+        kernel.operation.write_bytecode(out)
+      bc["custom_call_config"]["body"] = base64.b64encode(
+          out.getvalue()).decode()
+      op.attributes["backend_config"] = ir.StringAttr.get(
+          json.dumps(bc, separators=(",", ":")))
+    except Exception as e:  # pylint: disable=broad-except
```
Comment (skye) [inline jax/_src/cache_key.py]: This seems like it could mask actual bugs. Do we expect the above code to fail?

Note we already catch any cache failures to avoid the cache taking down jobs, e.g.: https://github.com/jax-ml/jax/blob/ebf39ff6839045d80b2e669102585916f8e128fb/jax/_src/compiler.py#L780-L785

File: jax/_src/cache_key.py [hunk 3]
```diff
@@ -190,6 +192,46 @@ def _update_bc_attribute(op: ir.Operation) -> ir.WalkResult:
...
+      logger.debug("Failed to strip Mosaic debug info: %s", e)
+    return ir.WalkResult.ADVANCE
+
+  m.operation.walk(_strip_kernel)
```
[the remark this answers is not in the document]
Comment (skye) [reply jax/_src/cache_key.py:234]: How annoying would it be to write a unit test for this? I don't think you should do what Gemini is suggesting :P Would it work to define two identical Pallas kernel twice in the test case, and verify the second definition gets a cache hit?

If it's too annoying we can commit as-is assuming you manually tested, but this is complex enough it'd be good to have test coverage.

**Review by skye — CHANGES_REQUESTED** [review]
Sorry for the delay and thanks for this fix! Overall looks good but a few comments.

File: jax/_src/cache_key.py (hunk 2 above)
Comment (colin-ant) [reply jax/_src/cache_key.py]: aah that's a good point, I'll take out the try/except in `cache_key.py` and let it bubble up to `compiler.py`

File: jax/_src/cache_key.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (colin-ant) [reply jax/_src/cache_key.py:234]: I can do that, I think it can be a simple kernel but with manually edited debug info to verify the second invocation hits the cache?

[the remark this answers is not in the document]
Comment (skye) [reply jax/_src/cache_key.py:234]: I'm imagining that if you define the same kernel twice in a row, they should have identical MLIR except different line numbers. Hopefully the debug info includes line numbers? But I could be misunderstanding how this works. I'm not sure if kernels have names, maybe the names would have to be different and make this not work.

I'm not sure, but manually editing the debug info could be so invasive that it's hard to verify the test is correct, or the test could be brittle. But it would probably be better than nothing. If you wanna try go for it, but I'm also okay skipping (if above simple idea doesn't work either).

File: jax/_src/cache_key.py (hunk 2 above)
Comment (colin-ant) [reply jax/_src/cache_key.py]: fixed

File: jax/_src/cache_key.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (colin-ant) [reply jax/_src/cache_key.py:234]: Defining the same kernel twice in nested scopes (it looks like `__name__` ends up as `kernel_name` in the IR, not debug info, so it has to match) + different line numbers worked. I don't think this version is too brittle, but happy to drop the test if it is.

File: tests/cache_key_test.py [hunk 4]
```diff
@@ -289,6 +290,38 @@ def test_identical_computations_different_metadata(self, include_metadata):
       key2 = cache_key.get(computation2, devices, compile_options, backend)
     self.assertEqual(include_metadata, key1 != key2)
 
+  @parameterized.parameters([False, True])
+  @jtu.thread_unsafe_test()  # env vars are not thread-safe
```
Comment (skye) [inline tests/cache_key_test.py]: Is this test using env vars? (I see this is included in several tests but I think it's cargo cult copy/paste :P)

File: tests/cache_key_test.py [hunk 5]
```diff
@@ -289,6 +290,38 @@ def test_identical_computations_different_metadata(self, include_metadata):
...
+        o_ref[...] = x_ref[...] + 1.0
+      return jax.jit(lambda x: pl.pallas_call(
+          kernel, out_shape=jax.ShapeDtypeStruct(x.shape, x.dtype))(x))
+
+    x = np.ones((128,), dtype=np.float32)
+    computation1 = make_caller_a().lower(x).compiler_ir()
+    computation2 = make_caller_b().lower(x).compiler_ir()
+
+    with config.compilation_cache_include_metadata_in_key(include_metadata):
+      hash1 = self.get_hashed_value(
```
Comment (skye) [inline tests/cache_key_test.py]: Can you use cache_key.get? Then you're more directly testing the cache_key behavior. I see test_identical_computations_different_metadata does this.

**Review by skye — APPROVED** [review]
LGTM! A few small comments on the test. Please respond and then we can get this in!

**Comment by skye** [comment]
FYI today is my last day at Google :open_mouth: 

This change is ready to go in once the last comments are addressed (if needed) and the presubmits pass, including the CLA. @jakevdp has agreed to help get this in, thanks!

File: tests/cache_key_test.py (hunk 4 above)
Comment (colin-ant) [reply tests/cache_key_test.py]: yes I was cargo culting it 😂

**Comment by colin-ant** [comment]
@jakevdp this should be good to go now, addressed @skye's review comments

File: tests/cache_key_test.py (hunk 5 above)
Comment (colin-ant) [reply tests/cache_key_test.py]: fixed

**Comment by colin-ant** [comment]
The previous CI failures should be fixed by the `main` merge above

**Comment by skye-ant** [comment]
@jakevdp can we get this in plz + ty <33
